### DISCUSSION TITLE
v2: Swapping out Provider stores with `useMemo`

### DISCUSSION THREAD

**Discussion by probeiuscorp** [pmndrs/jotai#1700] [Q&A]
Swapping out Provider stores with `useMemo` in the v2 API seems to be an incredibly powerful and simple way to initialize atoms. Unlike [using refs to initialize stores](https://github.com/pmndrs/jotai/discussions/1514#discussioncomment-4471595), it is implicitly typed (so you don't have to `useRef<Store>`) and **works even when the component is already mounted.**

```js
function useInitializedStore(initialize, deps = []) {
    return useMemo(() => {
        const store = createStore()
        initialize(store)
        return store
    }, deps)
}

const atomToInitialize = atom()
export function App({ userId }) {
    const store = useInitializedStore(store => {
        store.set(atomToInitialize, userId)
    }, [userId])

    return (
        <Provider store={store}>
            ...
        </Provider>
    )
}
```
[full working example](https://codesandbox.io/s/brave-dream-wf7vcx?file=/src/App.tsx)

I couldn't find anything mentioning this pattern on Issues or in the documentation. To my knowledge swapping out stores doesn't cause any async problems (see example) and it beautifully sidesteps the updating-on-render problem that keeps `useRef` and `useHydrateAtoms` mount-only. Is there a reason not to use this?

**Comment by dai-shi** [comment]
Yeah, we wish we could use this pattern with `useMemo`.
What's blocking is the official docs https://reactjs.org/docs/hooks-reference.html#usememo saying:

> You may rely on useMemo as a performance optimization, not as a semantic guarantee. In the future, React may choose to “forget” some previously memoized values and recalculate them on next render, e.g. to free memory for offscreen components. Write your code so that it still works without useMemo — and then add it to optimize performance.

This is not only with jotai. The whole React community is trying to avoid `useMemo` as a semantic guarantee.

**Reply by probeiuscorp** [reply]
:(

Can we not make a `useMemo` lookalike with a semantic guarantee by manually diffing the deps and just storing it in a `useRef`?

**Reply by dai-shi** [reply]
To make it concurrent friendly, something like this:
```js
const useSafeMemo = (fn, deps) => {
  const [[currValue, currDeps], set] = useState(() => [fn(), deps]);
  let value = currValue;
  const changed = deps.length !== currDeps.length || deps.some((dep, i) => dep !== currDeps[i]);
  if (changed) {
    value = fn();
    set([value, deps]);
  }
  return value;
};
```
